homeassistant/components/nibe_heatpump/__init__.py

```diff
@@ -0,0 +1,226 @@
+"""The Nibe Heat Pump integration."""
+from __future__ import annotations
+
+from datetime import timedelta
+
+from nibe.coil import Coil
+from nibe.connection import Connection
+from nibe.connection.nibegw import NibeGW
+from nibe.exceptions import CoilNotFoundException, CoilReadException
+from nibe.heatpump import HeatPump, Model
+from tenacity import RetryError, retry, retry_if_exception_type, stop_after_attempt
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_IP_ADDRESS, CONF_MODEL, Platform
+from homeassistant.core import CALLBACK_TYPE, HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady, HomeAssistantError
+from homeassistant.helpers import device_registry as dr
+from homeassistant.helpers.entity import DeviceInfo, async_generate_entity_id
+from homeassistant.helpers.update_coordinator import (
+    CoordinatorEntity,
+    DataUpdateCoordinator,
+)
+
+from .const import (
+    CONF_CONNECTION_TYPE,
+    CONF_CONNECTION_TYPE_NIBEGW,
+    CONF_LISTENING_PORT,
+    CONF_REMOTE_READ_PORT,
+    CONF_REMOTE_WRITE_PORT,
+    CONF_WORD_SWAP,
+    DOMAIN,
+    LOGGER,
+)
+
+PLATFORMS: list[Platform] = [Platform.SENSOR]
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Nibe Heat Pump from a config entry."""
+
+    heatpump = HeatPump(Model[entry.data[CONF_MODEL]])
+    heatpump.word_swap = entry.data[CONF_WORD_SWAP]
+    heatpump.initialize()
+
+    connection_type = entry.data[CONF_CONNECTION_TYPE]
+
+    if connection_type == CONF_CONNECTION_TYPE_NIBEGW:
+        connection = NibeGW(
+            heatpump,
+            entry.data[CONF_IP_ADDRESS],
+            entry.data[CONF_REMOTE_READ_PORT],
+            entry.data[CONF_REMOTE_WRITE_PORT],
+            listening_port=entry.data[CONF_LISTENING_PORT],
+        )
+    else:
+        raise HomeAssistantError(f"Connection type {connection_type} is not supported.")
+
+    await connection.start()
+    coordinator = Coordinator(hass, heatpump, connection)
+
+    data = hass.data.setdefault(DOMAIN, {})
+    data[entry.entry_id] = coordinator
+
+    try:
+        await coordinator.async_config_entry_first_refresh()
+    except ConfigEntryNotReady:
+        await connection.stop()
+        raise
+
+    reg = dr.async_get(hass)
+    reg.async_get_or_create(
+        config_entry_id=entry.entry_id,
+        identifiers={(DOMAIN, entry.unique_id or entry.entry_id)},
+        manufacturer="NIBE Energy Systems",
+        model=heatpump.model.name,
+        name=heatpump.model.name,
+    )
+
+    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
+
+    # Trigger a refresh again now that all platforms have registered
+    hass.async_create_task(coordinator.async_refresh())
```

Comment:
I guess this is done as we only fetch data from the coils of the connected listeners. But then why do we have a call to `coordinator.async_config_entry_first_refresh` above?

Reply:
Before the entities have registered with the coordinator, it will not fetch any data at all. It's the enabled/registered entities that control what data is fetch from the pump.

Reply:
Right, so then `coordinator.async_config_entry_first_refresh` seems to be a no-op.

Reply:
Indeed so. Dropped for now, will likely need at some point, but not at this moment. https://github.com/home-assistant/core/pull/78931/commits/2e701d37b1cfec1fb056fedf0e670abf33f75a8d